modules/gateway/peersmanager.go

```diff
@@ -73,6 +74,29 @@ func (g *Gateway) permanentPeerManager(closedChan chan struct{}) {
 	connectionLimiterChan := make(chan struct{}, maxConcurrentOutboundPeerRequests)
 
 	g.log.Debugln("INFO: [PPM] Permanent peer manager has started")
+
+	// Collect a list of peers that we have previously connected to.
+	g.mu.RLock()
+	var previousOutboundPeers []modules.NetAddress
+	for _, node := range g.nodes {
+		if node.WasOutboundPeer {
+			previousOutboundPeers = append(previousOutboundPeers, node.NetAddress)
+		}
+	}
+	g.mu.RUnlock()
+
+	// Construct a generator to return the next node considered for
+	// connection. The generator will first draw randomly from
+	// previousOutboundPeers, and then fallback to random node selection.
+	perm := fastrand.Perm(len(previousOutboundPeers))
+	nextNode := func() (addr modules.NetAddress, err error) {
+		if len(perm) > 0 {
+			addr, perm = previousOutboundPeers[perm[0]], perm[1:]
+			return addr, nil
+		}
+		return g.randomNode()
+	}
```

Comment:
this is a little fancy, but the alternative would be to either duplicate the following `for` loop or mangle it to accommodate both sources.